File: mobile/ios/Accord/Services/APIService.swift
// APIService.swift — REST API client for the Accord relay

import Foundation

/// REST client for relay HTTP endpoints (user registration, key upload, node management, etc.)
@Observable
final class APIService {
    private let baseURL: String
    private let session = URLSession.shared
    private var authToken: String?

    init(baseURL: String) {
        self.baseURL = baseURL
    }

    func setAuthToken(_ token: String) {
        self.authToken = token
    }

    // MARK: - Generic Request

    private func request<T: Decodable>(_ method: String, path: String, body: (any Encodable)? = nil) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw URLError(.badURL)
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            req.httpBody = try JSONEncoder().encode(body)
        }
        let (data, _) = try await session.data(for: req)
        return try JSONDecoder().decode(T.self, from: data)
    }

    // MARK: - Identity

    /// Upload our publishable key bundle to the relay.
    func uploadKeyBundle(_ bundle: Data) async throws {
        // TODO: POST /keys with bundle data
    }

    /// Fetch a peer's key bundle for initiating a session.
    func fetchKeyBundle(userId: String) async throws -> Data {
        // TODO: GET /keys/{userId}
        fatalError("Not implemented")
    }

    // MARK: - Nodes

    func fetchNodes() async throws -> [Node] {
        // TODO: GET /nodes
        return []
    }

    func fetchChannels(nodeId: String) async throws -> [Channel] {
        // TODO: GET /nodes/{nodeId}/channels
        return []
    }

    // MARK: - Messages

    func fetchMessages(channelId: String, before: String? = nil, limit: Int = 50) async throws -> [Message] {
        // TODO: GET /channels/{channelId}/messages
        return []
    }

    func sendMessage(channelId: String, encryptedContent: Data) async throws {
        // TODO: POST /channels/{channelId}/messages
    }

    // MARK: - DMs

    func fetchDMConversations() async throws -> [(User, Message?)] {
        // TODO: GET /dms
        return []
    }

    // MARK: - Users

    func fetchUser(id: String) async throws -> User {
        // TODO: GET /users/{id}
        fatalError("Not implemented")
    }
}

File: mobile/android/app/src/main/java/com/accord/data/service/ApiService.kt
package com.accord.data.service

import com.accord.data.model.Channel
import com.accord.data.model.Message
import com.accord.data.model.Node
import com.accord.data.model.User
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.http.*

/**
 * Retrofit REST API interface for Accord relay HTTP endpoints.
 */
interface AccordApi {
    // ── Nodes ────────────────────────────────────────────────────────
    @GET("nodes")
    suspend fun getNodes(): List<Node>

    @GET("nodes/{nodeId}/channels")
    suspend fun getChannels(@Path("nodeId") nodeId: String): List<Channel>

    // ── Messages ─────────────────────────────────────────────────────
    @GET("channels/{channelId}/messages")
    suspend fun getMessages(
        @Path("channelId") channelId: String,
        @Query("before") before: String? = null,
        @Query("limit") limit: Int = 50,
    ): List<Message>

    @POST("channels/{channelId}/messages")
    suspend fun sendMessage(
        @Path("channelId") channelId: String,
        @Body body: Map<String, String>,
    ): Message

    // ── Users / Keys ─────────────────────────────────────────────────
    @GET("users/{userId}")
    suspend fun getUser(@Path("userId") userId: String): User

    @GET("users/{userId}/bundle")
    suspend fun getPreKeyBundle(@Path("userId") userId: String): Map<String, String>

    @PUT("users/me/bundle")
    suspend fun uploadPreKeyBundle(@Body bundle: Map<String, String>)

    // ── DMs ──────────────────────────────────────────────────────────
    @GET("dms")
    suspend fun getDMChannels(): List<Channel>

    @POST("dms")
    suspend fun createDM(@Body body: Map<String, String>): Channel
}

/**
 * Factory for creating the API service with a given relay base URL.
 */
object ApiService {
    fun create(baseUrl: String): AccordApi {
        return Retrofit.Builder()
            .baseUrl(baseUrl.trimEnd('/') + "/")
            .addConverterFactory(MoshiConverterFactory.create())
            .build()
            .create(AccordApi::class.java)
    }

    // TODO: Add OkHttp interceptor for auth token injection
    // TODO: Add certificate pinning for relay connections
}
